crates/context_aware_config/src/api/custom_types/mod.rs

```diff
@@ -0,0 +1,157 @@
+    })?;
+    if !regex.is_match(type_name.as_str()) {
+        return Err(bad_argument!(
+            "The type name {} is invalid, it should obey the regex {}",
+            type_name,
+            TYPE_NAME_REGEX
+        ));
+    }
+    let type_template = diesel::insert_into(jsonschema_types::table)
+        .values((
+            jsonschema_types::type_schema.eq(request.type_schema.clone()),
+            jsonschema_types::type_name.eq(request.type_name.clone()),
+            jsonschema_types::created_by.eq(user.email.clone()),
+        ))
+        .get_result::<JsonSchemaTypes>(&mut conn)
+        .map_err(|err| {
+            log::error!("failed to insert custom type with error: {}", err);
+            db_error!(err)
+        })?;
+    Ok(HttpResponse::Ok().json(type_template))
+}
+
+#[post("")]
+async fn update_type(
+    request: Json<TypeTemplateRequest>,
+    db_conn: DbConnection,
+    user: User,
+) -> superposition::Result<HttpResponse> {
+    let DbConnection(mut conn) = db_conn;
+    let _ = JSONSchema::compile(&request.type_schema).map_err(|err| {
+        log::error!(
+            "Invalid jsonschema sent in the request, schema: {:?} error: {}",
+            request.type_schema,
+            err
+        );
+        bad_argument!(
+            "Invalid jsonschema sent in the request, validation error is: {}",
+            err.to_string()
+        )
+    })?;
+    let type_name = request.type_name.clone();
+    let regex = Regex::new(TYPE_NAME_REGEX).map_err(|err| {
+        unexpected_error!("could not parse regex due to: {}", err.to_string())
+    })?;
+    if !regex.is_match(type_name.as_str()) {
+        return Err(bad_argument!(
+            "The type name {} is invalid, it should obey the regex {}",
+            type_name,
+            TYPE_NAME_REGEX
+        ));
+    }
+    let timestamp = Utc::now().naive_utc();
+    let updated_type = diesel::update(jsonschema_types::table)
+        .filter(jsonschema_types::type_name.eq(request.type_name.clone()))
+        .set((
+            jsonschema_types::type_schema.eq(request.type_schema.clone()),
+            jsonschema_types::created_by.eq(user.email),
+            jsonschema_types::last_modified.eq(timestamp),
+        ))
+        .get_result::<JsonSchemaTypes>(&mut conn)
+        .map_err(|err| {
+            log::error!("failed to insert custom type with error: {}", err);
+            db_error!(err)
+        })?;
+    Ok(HttpResponse::Ok().json(updated_type))
+}
+
+#[delete("/{type_name}")]
+async fn delete_type(
+    path: Path<String>,
+    db_conn: DbConnection,
+) -> superposition::Result<HttpResponse> {
+    let DbConnection(mut conn) = db_conn;
+    let type_name = path.into_inner();
+    let deleted_type =
+        diesel::delete(dsl::jsonschema_types.filter(dsl::type_name.eq(type_name)))
+            .get_result::<JsonSchemaTypes>(&mut conn)?;
+    Ok(HttpResponse::Ok().json(deleted_type))
+}
+
+#[get("/{page}/{count}")]
```

Comment:
Having the page and count in the URL makes it mandatory for the user to opt into pagination, i think this is not needed it can be in `query params`. 
For the cases like where we want to list all the types, for example the dimension_form or the default_config_form, we won't have a paginated view, we need all the types in one go, so that they can be displayed in the dropdown

Reply:
If they are query params, you have to assume a default anyway. This is a better way of organizing list APIs, they should all be paginated. The changes need to propagate to dimensions and default config as well so that the frontend doesn't have to guess the page length and count

Reply:
i too think these should be query_params instead of having it in URL.
https://docs.github.com/en/rest/using-the-rest-api/using-pagination-in-the-rest-api?apiVersion=2022-11-28
https://medium.com/@premdattan/pagination-and-filter-url-strategy-b67d690120ae

Reply:
Leaving it as is, we can sync up on list API formats later